### PR TITLE
Support of integral and enum template parameters.

### PR DESCRIPTION
**Pull request by aneshlya**
This PR requires https://github.com/ispc/ispc/pull/2789.
Introduces support of integral and enum template parameters.
The docs will be added after initial review.

### CODE REVIEW

File: tests/lit-tests/func_template_nontype_5.ispc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    return k;
+}
+
+uniform int bar_uniform(uniform int k) {
+    return enum_foo<uniform int, TWO>(k);
+}
+
+int bar_varying(uniform int k) {
+    return enum_foo<varying int, TWO>(k);
+}
```
Comment (nurmukhametov) [inline tests/lit-tests/func_template_nontype_5.ispc:38]: I would add test like this here:
```ispc
template<Foo N>
uniform int func() {
    switch (N) {
        case ZERO:
            return 0;
        case ONE:
            return 1;
        case TWO:
            return 2;
    }
    return 42;
}

uniform int func_1() { return func<ONE>(); } // ideally to check that it is just ret i32 1
```

Comment (aneshlya) [reply tests/lit-tests/func_template_nontype_5.ispc:49]: Added.

Comment (nurmukhametov) [reply tests/lit-tests/func_template_nontype_5.ispc:49]: I see that templated function are not inlined by some reason:
```llvm-ir
; Function Attrs: nounwind
define i32 @bar_uniform___uni(i32 %k, <8 x i32> %__mask) local_unnamed_addr #1 {
allocas:
  %calltmp = tail call i32 @enum_foo___uniCunenum_5B_Foo_5D_TWO___uni(i32 %k, <8 x i32> %__mask)
  ret i32 %calltmp
}

; Function Attrs: noinline nounwind
define linkonce_odr i32 @enum_foo___uniCunenum_5B_Foo_5D_TWO___uni(i32 %k, <8 x i32> %__mask) local_unnamed_addr #2 {
allocas:
  ret i32 2
}
```

Comment (aneshlya) [reply tests/lit-tests/func_template_nontype_5.ispc:49]: Yes, they are declared with `noinline` keyword.

Comment (nurmukhametov) [reply tests/lit-tests/func_template_nontype_5.ispc:49]: Ah, I see now. But why do we need them? My initial point was to test exactly this:
```ispc
// CHECK-LABEL: @bar_uniform___uni
// CHECK-NEXT: allocas:
// CHECK-NEXT: ret i32 2
uniform int bar_uniform(uniform int k) {
    return enum_foo<uniform int, TWO>(k);
}
```

**Review by dbabokin — COMMENTED** [review]
I see a problem using template arguments:

```
> cat template.ispc
template <int I> uniform int foo1() {
  return I + 1;
}

template <int J> uniform int foo2() {
  return foo1<J>() + 1;
}

uniform int bar() {
  return foo2<1>();
}
(.venv_shark) DMITRYs-MacBook-Pro-2:~/ispc/dbabokin_ispc/wt/templ-nontype4/build (templ-nontype4)> bin/ispc template.ispc
Warning: No output file or header file name specified. Program will be compiled and warnings/errors will be issued, but no output will be generated.
Warning: No --target specified on command-line. Using default system target "neon-i32x4".
template.ispc:2:10: Error: Assertion failed (/Users/dbabokin/ispc/dbabokin_ispc/wt/templ-nontype4/src/expr.cpp:2114): "m->errorCount > 0".
  return I + 1;
         ^^^^^

***
*** Please file a bug report at https://github.com/ispc/ispc/issues
*** (Including as much information as you can about how to reproduce this error).
*** You have apparently encountered a bug in the compiler that we'd like to fix!
***
```

**Comment by aneshlya** [comment]
> I see a problem using template arguments:
> 
> ```
> > cat template.ispc
> template <int I> uniform int foo1() {
>   return I + 1;
> }
> 
> template <int J> uniform int foo2() {
>   return foo1<J>() + 1;
> }
> 
> uniform int bar() {
>   return foo2<1>();
> }
> (.venv_shark) DMITRYs-MacBook-Pro-2:~/ispc/dbabokin_ispc/wt/templ-nontype4/build (templ-nontype4)> bin/ispc template.ispc
> Warning: No output file or header file name specified. Program will be compiled and warnings/errors will be issued, but no output will be generated.
> Warning: No --target specified on command-line. Using default system target "neon-i32x4".
> template.ispc:2:10: Error: Assertion failed (/Users/dbabokin/ispc/dbabokin_ispc/wt/templ-nontype4/src/expr.cpp:2114): "m->errorCount > 0".
>   return I + 1;
>          ^^^^^
> 
> ***
> *** Please file a bug report at https://github.com/ispc/ispc/issues
> *** (Including as much information as you can about how to reproduce this error).
> *** You have apparently encountered a bug in the compiler that we'd like to fix!
> ***
> ```

It seems that this problem is not related to the changes introduced in this PR, it existed before: https://ispc.godbolt.org/z/fe9szMa4v

It happens when template parameter is not used in function arguments. You can see that it works when `foo2` has templated argument (https://ispc.godbolt.org/z/eK8j1bqcq). When function has templated arguments, we check it in `ResolveOverloads` https://github.com/ispc/ispc/blob/main/src/expr.cpp#L8900 and return earlier before doing `AddInstantiation`. If not, we continue executing the function and create a new instantiation, but it's too earlier for that.

File: docs/ispc.rst [hunk 2]
```diff
@@ -3907,6 +3907,36 @@ template function specializations is not yet supported. Consider the following e
...
+        return a * ab;
+      }
+
+      int qux() {
+        return baz<B>(3); // returns 6
+      }
+
+* Varying types are not allowed.
+* For integral types only constants can be used as non-type template arguments.
+  Constant expressions are not allowed.
```
[the remark this answers is not in the document]
Comment (aneshlya) [reply docs/ispc.rst]: Updated

File: docs/ispc.rst [hunk 3]
```diff
@@ -3907,6 +3907,36 @@ template function specializations is not yet supported. Consider the following e
       return a1 * a2;
     }
 
+For non-type template parameters, the following rules apply:
+
+* Uniform integral types (``int8``, ``uint8``, ``int16``, ``uint16``, ``int``, ``uint``, ``int64``, ``uint64``)
+  and enum types can be used as non-type template parameters. Unbound types are treated as uniform.
```
Comment (dbabokin) [inline docs/ispc.rst:3913]:
```suggestion
* Uniform integral types and enum types can be used as non-type template parameters. Unbound types are treated as uniform.
```

**Review by dbabokin — APPROVED** [review]
Looks good.

**Pull request merged**